Authenticator/Source/TableDiff.swift

```diff
@@ -31,92 +31,63 @@ protocol Identifiable {
+func changesFrom<T: Identifiable>(oldItems: [T], to newItems: [T]) -> [Change] {
+    return changesFrom(oldItems, to: newItems, hasSameIdentity: { $0.hasSameIdentity($1) },
+        isEqual: { (_, _) in false })
 }
 
-func changesFrom<T: Identifiable where T: Equatable>(oldArray: [T], to newArray: [T]) -> [Change] {
-    return changes(from: ArraySlice(oldArray), to: ArraySlice(newArray),
-        comparator: { $0.hasSameIdentity($1) }, equation: ==)
+func changesFrom<T: Identifiable where T: Equatable>(oldItems: [T], to newItems: [T]) -> [Change] {
+    return changesFrom(oldItems, to: newItems, hasSameIdentity: { $0.hasSameIdentity($1) },
+        isEqual: ==)
 }
 
-private func changes<Row>(from oldRows: ArraySlice<Row>, to newRows: ArraySlice<Row>,
-    comparator rowsHaveSameIdentity: (Row, Row) -> Bool, equation rowsAreEqual: (Row, Row) -> Bool)
-    -> [Change]
+// Diff algorithm from the Eugene Myers' paper "An O(ND) Difference Algorithm and Its Variations"
+private func changesFrom<T>(oldItems: [T], to newItems: [T], hasSameIdentity: (T, T) -> Bool,
+    isEqual: (T, T) -> Bool) -> [Change]
 {
-    // Work from the end to preserve earlier indices when recursing
-    switch (oldRows.last, newRows.last) {
-    case let (.Some(oldRow), .Some(newRow)):
-        if rowsHaveSameIdentity(oldRow, newRow) {
-            // The old and new rows have the same identity...
-            if rowsAreEqual(oldRow, newRow) {
-                // ...and are truly equal, so no change is needed.
-                return changes(from: oldRows.dropLast(), to: newRows.dropLast(),
-                    comparator: rowsHaveSameIdentity, equation: rowsAreEqual)
+    let MAX = oldItems.count + newItems.count
+    guard MAX > 0 else { return [] }
+    let numDiagonals = (2 * MAX) + 1
+    var V: [Int] = Array(count: numDiagonals, repeatedValue: 0)
+    var changesInDiagonal: [[Change]] = Array(count: numDiagonals, repeatedValue: [])
+    for D in 0...MAX {
+        for k in (-D).stride(through: D, by: 2) {
+            var x: Int
+            var changes: [Change]
+            if D == 0 {
+                x = 0
+                changes = []
+            } else
+            if k == -D || (k != D && V[k-1 + MAX] < V[k+1 + MAX]) {
+                x = V[k+1 + MAX]
+                changes = changesInDiagonal[k+1 + MAX]
+                changes = changes + [.Insert(index: x-k - 1)]
             } else {
-                // ...but differ in some way. This can be represented by an Update.
-                return changesWithUpdate(from: oldRows, to: newRows,
-                    comparator: rowsHaveSameIdentity, equation: rowsAreEqual)
+                x = V[k-1 + MAX] + 1
+                changes = changesInDiagonal[k-1 + MAX]
+                changes = changes + [.Delete(index: x - 1)]
+            }
+            var y = x - k
+            while x < oldItems.count && y < newItems.count && hasSameIdentity(oldItems[x], newItems[y]) {
```

Comment:
Line Length Violation: Line should be 100 characters or less: currently 105 characters (line_length)